client/beefy/src/import.rs

```diff
@@ -0,0 +1,211 @@
+pub struct BeefyBlockImport<Block: BlockT, Backend, Client, RuntimeApi, I> {
+	backend: Arc<Backend>,
+	client: Arc<Client>,
+	runtime: Arc<RuntimeApi>,
+	inner: I,
+	justification_sender: BeefySignedCommitmentSender<Block>,
+}
+
+impl<Block: BlockT, BE, Client, Runtime, I: Clone> Clone
+	for BeefyBlockImport<Block, BE, Client, Runtime, I>
+{
+	fn clone(&self) -> Self {
+		BeefyBlockImport {
+			backend: self.backend.clone(),
+			runtime: self.runtime.clone(),
+			client: self.client.clone(),
+			inner: self.inner.clone(),
+			justification_sender: self.justification_sender.clone(),
+		}
+	}
+}
+
+impl<Block: BlockT, BE, Client, Runtime, I> BeefyBlockImport<Block, BE, Client, Runtime, I> {
+	/// Create a new BeefyBlockImport.
+	pub fn new(
+		backend: Arc<BE>,
+		client: Arc<Client>,
+		runtime: Arc<Runtime>,
+		inner: I,
+		justification_sender: BeefySignedCommitmentSender<Block>,
+	) -> BeefyBlockImport<Block, BE, Client, Runtime, I> {
+		BeefyBlockImport { backend, client, runtime, inner, justification_sender }
+	}
+}
+
+impl<Block, BE, Client, Runtime, I> BeefyBlockImport<Block, BE, Client, Runtime, I>
+where
+	Block: BlockT,
+	BE: Backend<Block>,
+	Runtime: ProvideRuntimeApi<Block>,
+	Runtime::Api: BeefyApi<Block> + Send + Sync,
+{
+	fn decode_and_verify(
+		&self,
+		encoded: &EncodedJustification,
+		number: NumberFor<Block>,
+		hash: <Block as BlockT>::Hash,
+	) -> Result<VersionedFinalityProof<NumberFor<Block>, Signature>, ConsensusError> {
+		let block_id = BlockId::hash(hash);
+		let validator_set = self
+			.runtime
+			.runtime_api()
+			.validator_set(&block_id)
+			.map_err(|e| ConsensusError::ClientImport(e.to_string()))?
+			.ok_or_else(|| ConsensusError::ClientImport("Unknown validator set".to_string()))?;
+
+		decode_and_verify_commitment::<Block>(&encoded[..], number, &validator_set)
+	}
+
+	/// Import BEEFY justification: Send it to worker for processing and also append it to backend.
+	///
+	/// This function assumes:
+	/// - `justification` is verified and valid,
+	/// - the block referred by `justification` has been imported _and_ finalized.
+	fn import_beefy_justification_unchecked(
+		&self,
+		number: NumberFor<Block>,
+		justification: VersionedFinalityProof<NumberFor<Block>, Signature>,
+	) {
+		// Append the justification to the block in the backend.
+		if let Err(e) = self.backend.append_justification(
+			BlockId::Number(number),
+			(BEEFY_ENGINE_ID, justification.encode()),
+		) {
+			error!(target: "beefy", "🥩 Error {:?} on appending justification: {:?}", e, justification);
+		}
+		// Send the justification to the BEEFY voter for processing.
+		match justification {
+			// TODO #11838: Should not unpack, these channels should also use
+			// `VersionedFinalityProof`.
+			VersionedFinalityProof::V1(signed_commitment) => self
+				.justification_sender
+				.notify(|| Ok::<_, ()>(signed_commitment))
+				.expect("forwards closure result; the closure always returns Ok; qed."),
```

Comment:
This isn't "panic-proof" but it shouldn't cause any issues. Worst case the process crashes and we never process this justification from the worker-side, but once we restart the process the worker should try and request a justification for mandatory blocks from sync anyway (in a future PR).